File: 05_speedup/problems/01_matmul/matmul.cpp
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <omp.h>

/* 密行列積 C = A * B (いずれも n x n).
   行列は 1 次元配列に格納する (A[i*n+j] が A の (i,j) 要素).
   浮動小数点演算は乗算 + 加算が n 回 / 要素なので, 全体で 2 n^3 flops. */

int main(int argc, char ** argv) {
  long n = (1 < argc ? atol(argv[1]) : 1000);
  double * A = (double *)malloc(sizeof(double) * n * n);
  double * B = (double *)malloc(sizeof(double) * n * n);
  double * C = (double *)malloc(sizeof(double) * n * n);
  assert(A && B && C);
  /* A も B も全要素 1.0 に初期化 -> C[i][j] = n になるはず (検算しやすい) */
  for (long i = 0; i < n * n; i++) { A[i] = 1.0; B[i] = 1.0; C[i] = 0.0; }
  printf("n = %ld\n", n);
  /* 計測開始 */
  double t0 = omp_get_wtime();
  /* 計算本体: 3 重ループ. C[i][j] += A[i][k] * B[k][j] */
  // TODO: いちばん外側の i ループの直前に #pragma omp parallel for を1行追加し, 行ごとに並列化せよ.
  for (long i = 0; i < n; i++) {
    for (long j = 0; j < n; j++) {
      double s = 0.0;
      for (long k = 0; k < n; k++) {
        s += A[i * n + k] * B[k * n + j];
      }
      C[i * n + j] = s;
    }
  }
  /* 計測終了 */
  double t1 = omp_get_wtime();
  double dt = t1 - t0;          /* sec */
  /* 検算: 全要素が n のはずなので, 総和は n * n * n */
  double checksum = 0.0;
  for (long i = 0; i < n * n; i++) checksum += C[i];
  double expected = (double)n * (double)n * (double)n;
  printf("checksum   : %.0f (expected %.0f) -> %s\n",
         checksum, expected, (checksum == expected ? "OK" : "NG"));
  double flops = 2. * (double)n * (double)n * (double)n;
  printf("elapsed    : %7.3f  sec\n", dt);
  printf("flops      : %.2e\n", flops);
  printf("%.3f GFLOPS\n", flops / dt * 1e-9);
  return 0;
}

File: 05_speedup/problems/01_matmul/matmul.f90
program matmul_speedup
  use omp_lib
  ! 密行列積 C = A * B (いずれも n x n).
  ! 行列は 1 次元配列に格納する (A(i*n+j+1) が A の (i,j) 要素, i,j は 0 始まり).
  ! 浮動小数点演算は乗算 + 加算が n 回 / 要素なので, 全体で 2 n^3 flops.
  character(len=32) :: arg
  integer(8) :: n, i, j, k
  real(8), allocatable :: A(:), B(:), C(:)
  real(8) :: s, t0, t1, dt, flops, checksum, expected
  n = 1000
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) n
  end if
  allocate(A(n * n), B(n * n), C(n * n))
  ! A も B も全要素 1.0 に初期化 -> C(i,j) = n になるはず (検算しやすい)
  A = 1.0d0; B = 1.0d0; C = 0.0d0
  print "(a,i0)", "n = ", n
  ! 計測開始
  t0 = omp_get_wtime()
  ! 計算本体: 3 重ループ. C[i][j] += A[i][k] * B[k][j]
  ! TODO: いちばん外側の i ループを !$omp parallel do ... !$omp end parallel do で囲み, 行ごとに並列化せよ.
  do i = 0, n - 1
     do j = 0, n - 1
        s = 0.0d0
        do k = 0, n - 1
           s = s + A(i * n + k + 1) * B(k * n + j + 1)
        end do
        C(i * n + j + 1) = s
     end do
  end do
  ! TODO: 上で始めた parallel do 領域を閉じる (!$omp end parallel do).
  ! 計測終了
  t1 = omp_get_wtime()
  dt = t1 - t0                  ! sec
  ! 検算: 全要素が n のはずなので, 総和は n * n * n
  checksum = sum(C)
  expected = real(n, 8) * real(n, 8) * real(n, 8)
  if (checksum == expected) then
     print "(a,f0.0,a,f0.0,a)", "checksum   : ", checksum, " (expected ", expected, ") -> OK"
  else
     print "(a,f0.0,a,f0.0,a)", "checksum   : ", checksum, " (expected ", expected, ") -> NG"
  end if
  flops = 2.d0 * real(n, 8) * real(n, 8) * real(n, 8)
  print "(a,f7.3,a)", "elapsed    : ", dt, "  sec"
  print "(a,es9.2)",  "flops      : ", flops
  print "(f7.3,a)", flops / dt * 1e-9, " GFLOPS"
end program matmul_speedup
